doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java

```diff
@@ -130,133 +184,98 @@ public void parse( Reader source, Sink sink )
-        {
-            extensions.add( extension );
-        }
-
-        extensions.add( FlexmarkDoxiaExtension.create() );
-        flexmarkOptions.set( com.vladsch.flexmark.parser.Parser.EXTENSIONS, extensions );
-        flexmarkOptions.set( HtmlRenderer.HTML_BLOCK_OPEN_TAG_EOL, false );
-        flexmarkOptions.set( HtmlRenderer.HTML_BLOCK_CLOSE_TAG_EOL, false );
-        flexmarkOptions.set( HtmlRenderer.MAX_TRAILING_BLANK_LINES, -1 );
-
-        com.vladsch.flexmark.parser.Parser parser = com.vladsch.flexmark.parser.Parser.builder( flexmarkOptions )
-                .build();
-        HtmlRenderer renderer = HtmlRenderer.builder( flexmarkOptions )
-                                    .linkResolverFactory( new FlexmarkDoxiaLinkResolver.Factory() )
-                                    .build();
-
 
+        // Now, build the HTML document
         StringBuilder html = new StringBuilder( 1000 );
         html.append( "<html>" );
         html.append( "<head>" );
-        Pattern metadataPattern = Pattern.compile( MULTI_MARKDOWN_METADATA_SECTION, Pattern.MULTILINE );
-        Matcher metadataMatcher = metadataPattern.matcher( text );
+
+        // First, we interpret the "metadata" section of the document and add the corresponding HTML headers
+        Matcher metadataMatcher = METADATA_SECTION_PATTERN.matcher( text );
         boolean haveTitle = false;
         if ( metadataMatcher.find() )
         {
-            metadataPattern = Pattern.compile( MULTI_MARKDOWN_METADATA_ENTRY, Pattern.MULTILINE );
-            Matcher lineMatcher = metadataPattern.matcher( metadataMatcher.group( 1 ) );
-            boolean first = true;
-            while ( lineMatcher.find() )
+            Matcher entryMatcher = METADATA_ENTRY_PATTERN.matcher( metadataMatcher.group( 0 ) );
+            while ( entryMatcher.find() )
             {
-                String key = StringUtils.trimToEmpty( lineMatcher.group( 1 ) );
-                if ( first )
-                {
-                    boolean found = false;
-                    for ( String k : STANDARD_METADATA_KEYS )
-                    {
-                        if ( k.equalsIgnoreCase( key ) )
-                        {
-                            found = true;
-                            break;
-                        }
-                    }
-                    if ( !found )
-                    {
-                        break;
-                    }
-                    first = false;
-                }
-                String value = StringUtils.trimToEmpty( lineMatcher.group( 2 ) );
+                String key = entryMatcher.group( 1 );
+                String value = entryMatcher.group( 2 );
                 if ( "title".equalsIgnoreCase( key ) )
                 {
                     haveTitle = true;
                     html.append( "<title>" );
-                    html.append( StringEscapeUtils.escapeXml( value ) );
+                    html.append( HtmlTools.escapeHTML( value, false ) );
```

Comment:
Why is here a `false` and not just one arg like the rest of the calls?

Reply:
The only difference between *false* and *true* in *HtmlTools.escapeHTML()* is the handling of the apostrophe (should it be replaced with `&apos;` or not). As we're inside a `<title>` element, there is no need to replace the apostrophe.

However, on line 218 and 221, we're inside the attribute of an element, i.e. inside quotes, and that's why we need to escape the apostrophe here.

Note that the presence (and absence) of apostrophe in the title and metadatas is covered with unit tests (see *testMetadataSinkEvent()* in **MarkdownParserTest.java**)

Reply:
I am not really happy with that because we have an XHTML (HTML) sink and `false` means that it uses HTML entities which XML does not have.

Reply:
Happy New Year! 😉

I understand your concern, but I think that the parameter names are misleading.

More importantly, the *XhtmlBaseSink* **itself** uses `HtmlTools.escapeHTML(text, false)`:

* For example, all XHTML content is inserted with [XhtmlBaseSink.content()](https://github.com/apache/maven-doxia/blob/74c3e876b5c7a044f5a6b7ad7be6ea3ac10a6f49/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/XhtmlBaseSink.java#L2176)
* *XhtmlBaseSink.content()* calls [XhtmlBaseSink.escapeHTML()](https://github.com/apache/maven-doxia/blob/74c3e876b5c7a044f5a6b7ad7be6ea3ac10a6f49/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/XhtmlBaseSink.java#L2201)

and *XhtmlBaseSink.escapeHTML()* is simply:

```java
    /**
     * Forward to HtmlTools.escapeHTML( text ).
     *
     * @param text the String to escape, may be null
     * @return the text escaped, "" if null String input
     * @see org.apache.maven.doxia.util.HtmlTools#escapeHTML(String)
     */
    protected static String escapeHTML( String text )
    {
        return HtmlTools.escapeHTML( text, false );
    }
```

So, calling `HtmlTools.escapeHTML( text, false )` is really the only proper way to produce XHTML content that will be parsed with the XhtmlSink parser. And it is consistent with Doxia's own XhtmlSink.

Reply:
I am sick and tired of the codebase we have. Hacks after hacks with escaping rules w/o any consistency. BUT you are right, lets keep it consistent even if it is wrong.

Reply:
I hear you! I guess *escapeHTML()* should be renamed *escapeXHTML()*, as it's what it actually does. And the *xml* param should really be just about apostrophes. Anyway, that's how it is now...

Thank you for the review, @michael-o !

Reply:
Even more than that. HTML5 requires so little escaping because we use UTF-8 now. I see little reason to replace one or two byte characters in UTF-8 with 5 bytes of entities. Insane. This has been done 10-15 years ago, but is obsolete now.